src/Language/Github/Actions/UnstructuredMap.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- Module      : Language.Github.Actions.UnstructuredMap
+-- Description : Flexible value types for GitHub Actions YAML parsing
+-- Copyright   : (c) 2025 Bellroy Pty Ltd
+-- License     : BSD-3-Clause
+-- Maintainer  : Bellroy Tech Team <haskell@bellroy.com>
+--
+-- This module provides the 'UnstructuredValue' type for representing values
+-- that can be strings, numbers, or booleans in GitHub Actions YAML files.
+-- This is commonly needed when parsing @Map Text Text@ fields that GitHub
+-- Actions allows to have flexible typing.
+--
+-- GitHub Actions allows flexible typing in many contexts:
+-- * @retention-days: 1@ (number)
+-- * @retention-days: "1"@ (string)
+-- * @if: false@ (boolean)
+-- * @timeout-minutes: 30@ (number)
+-- * @working-directory: "./src"@ (string)
+--
+-- This type preserves the original YAML type during round-trip parsing,
+-- ensuring that numeric values remain numeric and strings remain strings.
+module Language.Github.Actions.UnstructuredMap
+  ( UnstructuredValue (..),
+    UnstructuredMap (..),
+    renderUnstructuredValue,
+    gen,
+  )
+where
+
+import Data.Aeson (FromJSON, ToJSON (..), Value (..))
+import qualified Data.Aeson as Aeson
+import Data.Map (Map)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import GHC.Generics (Generic)
+import Hedgehog (MonadGen)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+-- | A map that can have values of string, number, or boolean.
+--
+-- This type is designed to handle the flexible typing that GitHub Actions
+-- allows in YAML files.
+--
+-- The type preserves the original format during round-trip serialization,
+-- so numeric inputs remain numeric in the output YAML.
+data UnstructuredValue
```

Comment:
Can GHA YAML use nulls here, and if so, should we represent them?

Reply:
I wonder if `ScalarValue` would have been a more meaningful name here? "Unstructured" can mean different things to different people.

It's still debatable, of course, whether a `Null` value is "structured" or "scalar".